wgsl/index.bs

```diff
@@ -288,9 +288,10 @@ There are three kinds of errors, corresponding to the shader lifecycle:
     and other information available to the `createShaderModule` API method.
 
 * A <dfn export>pipeline-creation error</dfn>
-    is an error feasibly detectable at [=pipeline creation=] time.
+    is an error detectable at [=pipeline creation=] time.
     Detection must rely only on the [SHORTNAME] program source text
     and other information available to the particular pipeline creation API method.
+    A complex shader program may cause pipeline creation to fail [=spuriously=].
```

Comment:
Did we actually agree that this needs to be allowed? I'm trying to see what kind of actions the user would take if they get a "spurious" error.

Reply:
What else would we do when the driver fails compilation? Lose the device?

Reply:
The discussion in #1872 is unfinished. My TL;DR is that maybe the spurious failures can be divided in 2 groups:
  - driver bugs reported upstream, manifested as device loss on our side
  - out of memory issues

Otherwise, what's the goal for the spurious failure to be manifesting separately? Is it actionable?

Reply:
Pipeline creation might fail without being OOM per se, and without being something an IHV might consider to be a bug. Generally speaking we want to minimize when we force device loss. It's not super actionable, but the *rest* of the app will still run, rather force-crashing it.

Reply:
In https://github.com/gpuweb/gpuweb/issues/1872#issuecomment-875882952 I wrote that a conformant Vulkan implementation has essentially only one way to tell us that something went wrong at pipeline creation, and it's OOM. So why should we do anything different than just pass it up to the user as an OOM error?

Reply:
Is that true besides Vulkan?

Reply:
Same for D3D12, see [documentation](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12device-creategraphicspipelinestate) of `CreateGraphicsPipelineState`:
> This method returns E_OUTOFMEMORY if there is insufficient memory to create the pipeline state object.

Reply:
"insufficient memory to create the pipeline state object" isn't what I had in mind. I think we should lose the device in that case because it should be so rare and difficult to recover from that I don't think anyone is going to. (We can add an opt-in feature if people want to recover from that.)

I am still thinking about the "unable to allocate registers+workgroup memory" problem, which I don't consider to be an OOM.

Reply:
How are you going to know if it's specifically register memory, given that the native APIs don't tell you that?
As for workgroup, we should be able to provide a way that *guarantees* successful allocation, or otherwise the workgroup memory limit is useless.

Reply:
`E_OUTOFMEMORY` is not the only return code possible from `CreateGraphicsPipelineState`. What's more, the debug layer returns much more helpful messages which is what developers rely on to resolve any such error

Reply:
It's really hard to see what other error codes are possible from `CreateGraphicsPipelineState`, specifically when WebGPU guarantees that all the expectations of this functions are fulfilled (e.g. not passing in garbage pointers):
> Return value
>Type: HRESULT
> This method returns E_OUTOFMEMORY if there is insufficient memory to create the pipeline state object. See Direct3D 12 Return Codes for other possible return values.

Visiting the [Direct3D 12 Return Codes](https://docs.microsoft.com/en-us/windows/win32/direct3d12/d3d12-graphics-reference-returnvalues) doesn't help much either. `E_NOTIMPL`?